src/main/java/org/opensearch/knn/index/memory/SharedIndexStateManager.java

```diff
@@ -0,0 +1,145 @@
+     *
+     * @param indexAddress Address of index to initialize the shared state from
+     * @param knnEngine engine index belongs to
+     * @return ShareModelContext
+     */
+    public SharedIndexState get(long indexAddress, String modelId, KNNEngine knnEngine) {
+        this.readWriteLock.readLock().lock();
+        // This can be done safely with readLock because the ConcurrentHasMap.computeIfAbsent guarantees:
+        //
+        // "If the specified key is not already associated with a value, attempts to compute its value using the given
+        // mapping function and enters it into this map unless null. The entire method invocation is performed
+        // atomically, so the function is applied at most once per key. Some attempted update operations on this map
+        // by other threads may be blocked while computation is in progress, so the computation should be short and
+        // simple, and must not attempt to update any other mappings of this map."
+        //
+        // Ref:
+        // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-
+        SharedIndexStateEntry entry = sharedIndexStateCache.computeIfAbsent(modelId, m -> {
+            log.info("Loading entry to shared index state cache for model {}", modelId);
+            long sharedIndexStateAddress = JNIService.initSharedIndexState(indexAddress, knnEngine);
+            return new SharedIndexStateEntry(new SharedIndexState(sharedIndexStateAddress, modelId, knnEngine));
+        });
+        entry.incRef();
+        this.readWriteLock.readLock().unlock();
+        return entry.getSharedIndexState();
+    }
+
+    /**
+     * Indicate that the {@link SharedIndexState} is no longer being used. If nothing else is using it, it will be
+     * removed from the cache and evicted.
+     *
+     * After calling this method, {@link SharedIndexState} should no longer be used by calling thread.
+     *
+     * @param sharedIndexState to return to the system.
+     */
+    public void release(SharedIndexState sharedIndexState) {
+        this.readWriteLock.writeLock().lock();
+
+        if (!sharedIndexStateCache.containsKey(sharedIndexState.getModelId())) {
+            // This should not happen. Will log the error and return to prevent crash
+            log.error("Attempting to evict model from cache but it is not present: {}", sharedIndexState.getModelId());
+            this.readWriteLock.writeLock().unlock();
+            return;
+        }
+
+        long refCount = sharedIndexStateCache.get(sharedIndexState.getModelId()).decRef();
+        if (refCount <= 0) {
+            log.info("Evicting entry from shared index state cache for key {}", sharedIndexState.getModelId());
+            sharedIndexStateCache.remove(sharedIndexState.getModelId());
+            JNIService.freeSharedIndexState(sharedIndexState.getSharedIndexStateAddress(), sharedIndexState.getKnnEngine());
+        }
```

Comment:
With this logic it means that if some query is still using the SharedIndexState then we will not be able to release the memory. But if a user has done the clear cache api then in that case we will just skip over this release and return that we have cleared the cache?

Reply:
The release portion is tied to an IndexAllocation not the query. When the index allocation is closed, we will decrease the reference count to this. Once it goes to 0, we free. If an index is loaded and needs shared state, it will need to obtain read lock and so will be blocked. So, after this completes, it will have to load again.

On clear cache, all IndexAllocations will be closed leading to refcount going down.

Reply:
Okay, so this line will prevent https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/KNNWeight.java#L273 the scenario where query is happening and we cannot release the memory.

Reply:
yes thats correct